internal/dag/gatewayapi_processor_test.go

```diff
@@ -17,175 +17,190 @@ import (
+				fmt.Errorf("invalid hostname \"#projectcontour.io\": [a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"invalid listener hostname": {
+			listenerHost: listenerHostname("#projectcontour.io"),
+			hostnames:    []gatewayapi_v1alpha1.Hostname{},
+			want:         []string(nil),
+			wantError: []error{
+				fmt.Errorf("invalid hostname \"#projectcontour.io\": [a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"invalid listener wildcard hostname": {
+			listenerHost: listenerHostname("*.*.projectcontour.io"),
+			hostnames:    []gatewayapi_v1alpha1.Hostname{},
+			want:         []string(nil),
+			wantError: []error{
+				fmt.Errorf("invalid hostname \"*.*.projectcontour.io\": [a wildcard DNS-1123 subdomain must start with '*.', followed by a valid DNS subdomain, which must consist of lower case alphanumeric characters, '-' or '.' and end with an alphanumeric character (e.g. '*.example.com', regex used for validation is '\\*\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"),
+			},
+		},
+		"listener host & hostnames host do not exactly match": {
+			listenerHost: listenerHostname("listener.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"listener.projectcontour.io\" does not match hostnames hostname \"http.projectcontour.io\"")},
+		},
+		"listener host & hostnames host exactly match": {
+			listenerHost: listenerHostname("http.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      []string{"http.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & hostnames host match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.projectcontour.io",
+			},
+			want:      []string{"http.projectcontour.io"},
+			wantError: nil,
+		},
+		"listener host & hostnames host do not match wildcard host": {
+			listenerHost: listenerHostname("*.projectcontour.io"),
+			hostnames: []gatewayapi_v1alpha1.Hostname{
+				"http.example.com",
 			},
+			want:      nil,
+			wantError: []error{fmt.Errorf("gateway hostname \"*.projectcontour.io\" does not match hostnames hostname \"http.example.com\"")},
+		},
+		"listener host & wildcard hostnames host match": {
```

Comment:
is this allowed? this seems like the listener is more specific than the route, which could be not allowed?

Reply:
I'm not 100% sure, but the test is wrong. If the Gateway defines `hostname.projectcontour.io` and the route defines `*.projectcontour.io`, then I think `hostname.projectcontour.io` should end up being the match. Does that make sense?

Reply:
yeah i think if we do allow this, the route programmed should be for the more specific one

I guess the question is can a route have a wildcard on a gateway/listener with a specific match? `*.projectcontour.io` is a superset of `http.projectcontour.io` so it is a bit odd to me this could be valid? idk

Reply:
cc @projectcontour/maintainers if there are any other opinions on this

Reply:
@stevesloka IMO a wildcard hostname should not be allowed for a route bound to a gateway listener with a more precise hostname. The API does not specify this use case, so I suggest updating https://github.com/kubernetes-sigs/gateway-api/issues/654 to get input from the community.

Reply:
commented on the upstream issue, not sure what we want to do for now, restrict to something less permissive and expand or open it up now and restrict later